Add support for UGW 8.x (and OpenWRT in general)

vitalybu opened the pull request.

Commit:
tools: OpenWRT external toolchain cmake files and builder docker

CMake and Docker files for building prplMesh out-of-tree on OpenWRT based platforms

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
tools: Better support for external toolchains

Allow overriding the CMAKE_TOOLCHAIN_FILE flag with an environment
variable or from the command line.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tools: maptools: don't override STAGING_DIR

We set the STAGING_DIR environment variable from maptools because
openwrt toolchains give a warning if it is not set. However, if
STAGING_DIR is already set in the environment, we shouldn't override it.

Only set STAGING_DIR if not set, and add a comment to explain why.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
platform: bpl: remove support for older ugw (pre-8.x) versions

Remove legacy BPL implementation which was for pre-8.x UGW versions.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
cmake: add Find files to safe c libraries

Add Findsafec and Findslibc cmake files for finding libsafec (ugw) and
libslibc (rdkb) libraries. These libraries contain safer implementation
for standard C functions (e.g. strcpy(), sprintf()).

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
framework: bpl: Renamed RDKB BPL to UCI BPL

Renamed the RDKB BPL to UCI BPL as it'll be the default option for
OpenWRT based platforms. Added TARGET_PLATFORM_TYPE for specifying
sub-types (e.g. "ugw", "turris-omnia").

Updated the CMake file for UGW 8.x

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
common: bwl: Sync to latest DWPAL changes

Update the DWPAL HAL implementation with the latest changes in the DWPAL
library. This brings it up to date with D-WPAL commit
b1341a600d87bd808e160ef40ccd2cb51db18b2d

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
common: scripts: Add UGW support to the prplmesh_utils script

The BusyBox implementation of pgrep is quite limited. Some (older?)
versions do not display the command line arguments of the executable.
By matching the full command line (executable + arguments) we ensure
the correct behavior of pgrep.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
common: prplmesh_utils.sh: improve agent operational check

Since automatic channel selection is not necessarily performed, the
agent doesn't necessarily go to CSA state. In addition, there are also
error messages that contain the string "CSA".

Instead, check for 'goto STATE_OPERATIONAL'.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
agent: Fix compilation error for older GCC versions

Older GCC versions generate an error when trying to upcast SocketClient*
to Socket*. Since the controller socket will always be of the
SocketClient type, we can safely use SocketClient in this class.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
controller: Generalize RDKB readline workaround

It is no longer needed on UGW, so just match rdkb.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
controller: Add UGW support

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
controller: son_actions: Remove unused variables

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
controller: Allow nodes with unknown IP

Call handle_complete_connection for clients with unknown IP address
(as prplMesh doesn't require the clients to have an IP).

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
controller: master: compilation fix for RDKB

Some of the RDKB specific code was not updated after a TLV change. Do so
now.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
build: Remove STANDALONE build mode

The STANDALONE build mode is no longer used.
All platforms have a dedicated external toolchain cmake file that should
be used for out-of-tree builds. The path to this file can be set using
the CMAKE_TOOLCHAIN_FILE environment variable (which is also done by the
docker builders).

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

tomereli commented on tools/cmake/toolchain/common.cmake:
```diff
@@ -0,0 +1,19 @@
+  set(BoldMagenta "${Esc}[1;35m")
+  set(BoldCyan    "${Esc}[1;36m")
+  set(BoldWhite   "${Esc}[1;37m")
+endif()
```
Missing line break at EOF

vitalybu replied:
Fixed

tomereli commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+
+## General configuration
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR ARMV7 CACHE STRING "System Processor")
```
ARMV7? what?

vitalybu replied:
Leftover from the Turris Omnia file.
We don't use it anyway, so removing..

tomereli commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+set(CMAKE_LIBRARY_PATH /usr/lib /lib)
+
+## Library config
+set(BUILD_SHARED_LIBS ON CACHE BOOL "BUILD_SHARED_LIBS" FORCE)
```
How come we set both BUILD_SHARED and BUILD_STATIC?

vitalybu replied:
Why not?
Theoretically nothing prevents us from building both Shared and Static versions of the libs.
This cmake is based on a more generic version that supports various types of builds...

tomereli replied:
ok

tomereli commented on tools/docker/builder/OpenWRT.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+ENV PRPLMESH_PLATFORM_TOOLCHAIN_PREFIX  $PRPLMESH_PLATFORM_TOOLCHAIN_PREFIX
+
+# Default command
+ENTRYPOINT ["./prplMesh/tools/maptools.py", "build"]
```
Missing line break

vitalybu replied:
Fixed

tomereli commented on framework/cmake/Findsafec.cmake:
```diff
@@ -0,0 +1,35 @@
+    SAFEC_INCLUDE_DIRS
+)
+
+# message("SAFEC_LIBRARY: ${SAFEC_LIBRARY}")
```
remove comments

vitalybu replied:
Removed

tomereli commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -14,29 +14,64 @@ file(GLOB_RECURSE bpl_common_sources ${MODULE_PATH}/common/*.c*)
-    include_directories(${PLATFORM_INCLUDE_DIR})
+    # Platform definitions
+    add_definitions(
+        -DBPL_PASSIVE_MODE
```
We should make `BPL_PASSIVE_MODE` the only mode

tomereli replied:
Actually, this will be done as part of https://github.com/prplfoundation/prplMesh/commits/feature/remove-non-passive-mode-code

tomereli commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -82,6 +113,7 @@ endif()
 # Build the library
 add_library(${PROJECT_NAME} SHARED ${bpl_common_sources} ${bpl_platform_sources})
 set_target_properties(${PROJECT_NAME} PROPERTIES VERSION ${${PROJECT}_VERSION_STRING} SOVERSION ${${PROJECT}_VERSION_MAJOR})
+set_target_properties(${PROJECT_NAME} PROPERTIES LINK_FLAGS "-Wl,-z,defs")
```
Can you add a comment here saying what these properties are?

tomereli commented on framework/platform/bpl/uci/arp/bpl_arp.cpp:
```diff
@@ -16,7 +16,7 @@
 
 extern "C" {
-#define FEATURE_LQ_BRCTL_SHOWMACS_IFNAME
+#define FEATURE_BRCTL_SHOWMACS_IFNAME
```
(nitpick) this should be a separate commit

tomereli commented on common/CMakeLists.txt:
```diff
@@ -15,15 +15,21 @@ include(GNUInstallDirs)
 set(BEEROCKS_BRIDGE_IFACE "br-lan")
 set(BEEROCKS_PLATFORM_INIT "false")
-set(BEEROCKS_INSTALL_PATH "/opt/prplmesh")
+set(BEEROCKS_INSTALL_PATH "/opt/prplMesh")
```
I prefer to keep it as `prplmesh` lowercase (nitpick)

arnout replied:
Actually, there's no reason to install prplmesh in /opt at all. /opt is for stuff that doesn't want to integrate with the distro at all.

Not something that we should change right away, but we definitely don't want to make further bikeshedding changes to the installation path.

tomereli replied:
But for now, let us just keep it as `/opt/prplmesh`, no reason to change this in the first place unless it is to remove this completely as you suggested.

tomereli commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -12,46 +12,49 @@ file(GLOB_RECURSE bwl_common_sources ${MODULE_PATH}/common/*.c*)
 
-    set(BWL_TYPE "NL80211")
+        # dwpal search path
+        link_directories(${PLATFORM_STAGING_DIR}/opt/intel/lib)
```
Can we not use `link_directories` and instead add a `Finddwpal.cmake` and use `find_package` like we do for other libraries such as `safec`?

tomereli commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -12,46 +12,49 @@ file(GLOB_RECURSE bwl_common_sources ${MODULE_PATH}/common/*.c*)
+    set(BWL_TYPE "DWPAL")
+
+    # Extre directories
+    include_directories(${PLATFORM_INCLUDE_DIR}/wav-dpal)
```
same here

vitalybu replied:
Typo?
Fixed.

arnout replied:
No, the "same here" was about using find_package and adding a Findwavdpal.cmake.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -16,8 +16,16 @@
 #include <bcl/network/network_utils.h>
 #include <bcl/son/son_wireless_utils.h>
 #include <easylogging++.h>
+
```
Commit message should include the commit hash of the DWPAL library that was used.

vitalybu replied:
Are you sure it's the best way to keep compatibility?
Perhaps DWPAL maintainers should increase the library version when introducing breaking changes?

arnout replied:
Perhaps they should, but even then the commit message should mention the dwpal library version on which it is based. If we don't have a library version, use the commit hash.

vitalybu replied:
What's the standard here? Simply add something like:
DWPAL Library Hash: b1341a600d87bd808e160ef40ccd2cb51db18b2d?

tomereli replied:
Yes, this works

vitalybu replied:
Done

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -265,6 +267,6 @@ DATA_IFACE=enx00ee22aa28ef
+export LD_LIBRARY_PATH=@BEEROCKS_INSTALL_PATH@/lib:$LD_LIBRARY_PATH:@EXTRA_LD_PATH@
 
-main $@
+main $@
```
missing empty line at end of file

vitalybu replied:
Fixed

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -265,6 +267,6 @@ DATA_IFACE=enx00ee22aa28ef
 
 # Export MultiAP libs folder
-export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:@BEEROCKS_INSTALL_PATH@/lib
+export LD_LIBRARY_PATH=@BEEROCKS_INSTALL_PATH@/lib:$LD_LIBRARY_PATH:@EXTRA_LD_PATH@
```
Where is `EXTRA_LD_PATH` defined? not in this commit.

vitalybu replied:
It's in the CMakeLists.txt and is part of this commit...

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -51,23 +51,6 @@ endif()
 # Default Compiler flags
```
Please update the readme.md file with instructions on how to build out of tree

tomereli commented on tools/docker/builder/OpenWRT.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+###############################################################
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+# Copyright (c) 2019 Tomer Eliyahu (Intel)
```
Thanks, but you own the authorship for this file :)

tomereli commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -14,29 +14,64 @@ file(GLOB_RECURSE bpl_common_sources ${MODULE_PATH}/common/*.c*)
-    find_package(ubus REQUIRED)
-    list(APPEND BPL_LIBS ubus nl-3 nl-route-3 bridge ugwhelper msgapi cal capi blobmsg_json)
+        # Platform specific includes
+        include_directories(${PLATFORM_INCLUDE_DIR}/libnl3)
```
Here there are a bunch of non-modern cmake hacks for libbridge and libnl3 - we should replace them with find modules.

arnout replied:
We should actually create a Findlibnl3 for this, but OK, we already do it like this for bwl so that can be fixed later.

tomereli commented on common/beerocks/scripts/CMakeLists.txt:
```diff
@@ -2,9 +2,12 @@
+    set(PLATFORM_DB_PATH "/opt/prplMesh/share")
     set(PLATFORM_DB_PATH_TEMP "/tmp/beerocks")
+    if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
+        set(EXTRA_LD_PATH "/opt/intel/lib")
```
Oh yack. No need to do that - better add a fix in UGW and install what we need to /usr/lib. If you need help with getting this merged to UGW - I can help, just open a PR and send me the link.

vitalybu replied:
I don't think that you can *always* assume that all the dependencies will be in /usr/lib.
Don't see an issue with allowing additional library paths for specific platforms...

tomereli replied:
We should at least add a comment explaining why this is needed, e.g. which libraries are installed in non-conventional locations per platform.

tomereli replied:
I take it back - no need for this. It was a bug in UGW and I pushed a fix which will be merged soon.

rmelotte commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1686,10 +1686,10 @@ bool master_thread::handle_cmdu_1905_topology_notification(Socket *sd,
         if (database.get_node_ipv4(client_mac_str).empty()) {
             database.set_node_state(client_mac_str, beerocks::STATE_CONNECTED_IP_UNKNOWN);
             LOG(INFO) << "STATE_CONNECTED_IP_UNKNOWN for node mac " << client_mac_str;
-            return true;
```
UNKNOWN_IP states were removed from bml in c07a34, so I maybe it could be removed from the bcl as well?

vitalybu replied:
Agreed. No need for this state anymore.

arnout commented on tools/docker/builder/OpenWRT.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+# See LICENSE file for more details.
+###############################################################
+
+FROM ubuntu:16.04
```
Why 16.04?

vitalybu replied:
Legacy from the Turris Omnia builder docker (Turris 3.x tree required Ubuntu 16.04 for building successfully). Updated to 18.04

arnout commented on tools/docker/builder/OpenWRT.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+ENV PRPLMESH_PLATFORM_BASE_DIR          $PRPLMESH_PLATFORM_BASE_DIR
+ENV PRPLMESH_PLATFORM_BUILD_NAME        $PRPLMESH_PLATFORM_BUILD_NAME
+ENV PRPLMESH_PLATFORM_TOOLCHAIN         $PRPLMESH_PLATFORM_TOOLCHAIN
+ENV PRPLMESH_PLATFORM_TOOLCHAIN_PREFIX  $PRPLMESH_PLATFORM_TOOLCHAIN_PREFIX
```
From this PR it is not clear to me where you want to go with this Dockerfile, because it's not used anywhere. So for sure, it should be in a separate PR (no dependency with anything else in this PR).

I think the way we want to go is that we have a container which already has the openwrt SDK installed in a well-known location, so that the prplmesh build can use it directly and generate a package. This actually means we would need two dockerfiles: the first one to build the SDK, the second one to build prplmesh.

But that doesn't need to be done in one shot. What is important now is that at least there's a readme (or an update to the existing one) that explains how to use this docker image.

arnout commented on tools/cmake/toolchain/common.cmake:
```diff
@@ -0,0 +1,19 @@
+  set(BoldMagenta "${Esc}[1;35m")
+  set(BoldCyan    "${Esc}[1;36m")
+  set(BoldWhite   "${Esc}[1;37m")
+endif()
```
IMO there's no reason to put this stuff in the toolchainfile. We can just put it in `cmake/multiap-helpers.cmake`.

vitalybu replied:
Agreed. Moved to the helpers cmake.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,59 @@
+
+## General configuration
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_EXPORT_NO_PACKAGE_REGISTRY ON)
```
Why is this needed?

vitalybu replied:
It's not really needed, but for out-of-tree builds (which this file is used for) there's no need to update the package registry.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+
+# Platform specific
+set(TARGET_PLATFORM             "openwrt" CACHE STRING "Target Platform")
+set(TARGET_PLATFORM_TYPE        "$ENV{PRPLMESH_PLATFORM_TYPE}" CACHE STRING "Target Platform Type")
```
If you add a new cache variable, it should also be added to the top-level CMakeLists.txt.

vitalybu replied:
Why?

arnout replied:
Well, not necessarily in the top-level. My point is: a cache variable should not be introduced by the toolchainfile, only by the core CMakeLists.txt files. Remember that this openwrt.cmake is only included if you point cmake to it explicitly from the command line.

vitalybu replied:
Indeed. These variables should be set by either the toolchain file or the build system for in-tree builds (e.g. OpenWRT Makefile).
How do you propose adding them into the top-level cmake? Empty variables?

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+# Platform specific
+set(TARGET_PLATFORM             "openwrt" CACHE STRING "Target Platform")
+set(TARGET_PLATFORM_TYPE        "$ENV{PRPLMESH_PLATFORM_TYPE}" CACHE STRING "Target Platform Type")
+set(PLATFORM_BASE_DIR           "$ENV{PRPLMESH_PLATFORM_BASE_DIR}" CACHE STRING "Platform Base Directory")
+set(PLATFORM_BUILD_NAME         "$ENV{PRPLMESH_PLATFORM_BUILD_NAME}" CACHE STRING "Platform Build Name")
+set(PLATFORM_TOOLCHAIN          "$ENV{PRPLMESH_PLATFORM_TOOLCHAIN}" CACHE STRING "Platform Toolchain")
```
These three are only used a few lines below, so we can use `$ENV{...}` directly there.

If you do keep the set commands: they are internal variables, so CACHE should not be used.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+
+# Compiler Definitions
+add_definitions(
+    -DBEEROCKS_OPENWRT
```
This is not used anywhere, so there's no point adding it.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+# Compiler Flags and Search Paths
+set(CMAKE_C_FLAGS "--sysroot=${PLATFORM_STAGING_DIR}")
+set(CMAKE_CXX_FLAGS "--sysroot=${PLATFORM_STAGING_DIR}")
+set(CMAKE_LIBRARY_PATH /usr/lib /lib)
```
Why is this needed. And more importantly: is this correct? We're cross-compiling, so finding libraries in /usr/lib doesn't sound like a good idea...

vitalybu replied:
We also use "--sysroot" when invoking the compiler, so these paths are relative to the system root folder.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+
+## Library config
+set(BUILD_SHARED_LIBS ON CACHE BOOL "BUILD_SHARED_LIBS" FORCE)
+set(BUILD_STATIC_LIBS ON CACHE BOOL "BUILD_STATIC_LIBS" FORCE)
+set(BUILD_SHARED ON CACHE BOOL "BUILD_SHARED" FORCE)
+set(BUILD_STATIC ON CACHE BOOL "BUILD_STATIC" FORCE)
```
BUILD_SHARED_LIBS is an existing CMake variable that does have an effect (it will build all libraries as shared unless they are explicitly defined as STATIC). BUILD_SHARED, BUILD_STATIC and BUILD_STATIC_LIBS are not "well known" CMake variables, and they're not used anywhere in the prplMesh cmake files, so they're useless.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+set(CMAKE_LIBRARY_PATH /usr/lib /lib)
+
+## Library config
+set(BUILD_SHARED_LIBS ON CACHE BOOL "BUILD_SHARED_LIBS" FORCE)
```
I don't think we should do this.

Instead, in the top-level CMakeLists, we should add

```
option(BUILD_SHARED_LIBS "Build libraries as shared" ON)
```

We should *definitely* not force it on here.

arnout commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,60 @@
+set(CMAKE_CROSS_COMPILING ON)
+
+# Platform specific
+set(TARGET_PLATFORM             "openwrt" CACHE STRING "Target Platform")
```
I think it would be useful to split this file in two parts: one part which corresponds to the normal toolchainfile (i.e. sets compiler paths etc.), and another part which sets the prplmesh-specific options (e.g. this one).

But that's something that can be done later.

arnout commented on tools/commands/build.py:
```diff
@@ -192,9 +196,14 @@ def __init__(self, args):
+            if "CMAKE_TOOLCHAIN_FILE" in os.environ:
+                map_cmake_flags += ["CMAKE_TOOLCHAIN_FILE=" + os.path.abspath(os.environ.get('CMAKE_TOOLCHAIN_FILE'))]
+            else:
+                map_cmake_flags += ["CMAKE_TOOLCHAIN_FILE=external_toolchain.cmake"]
```
This can be simplified to:
```
map_cmake_flags.append("CMAKE_TOOLCHAIN_FILE=" +
           os.path.abspath(os.environ.get('CMAKE_TOOLCHAIN_FILE', 'external_toolchain.cmake')))
```

(nitpick)

arnout commented on tools/commands/build.py:
```diff
@@ -192,9 +196,14 @@ def __init__(self, args):
+                map_cmake_flags += ["CMAKE_TOOLCHAIN_FILE=external_toolchain.cmake"]
+        for name in _map_modules:
+            builder = cmakebuilder(name, modules_dir, build_dir, install_dir, args.cmake_verbose, args.make_verbose,
+                map_cmake_flags, args.generator)
```
This reverts a change that was just applied to master. Please don't do that :-)

arnout commented on tools/commands/build.py:
```diff
@@ -21,7 +21,11 @@ def __init__(self, name, modules_dir, build_dir, install_dir):
+        # If the STAGING_DIR environment variable is not set, 
+        # use an empty variable for suppressing compiler warnings
+        if not "STAGING_DIR" in os.environ:
+            self.env["STAGING_DIR"] = ""
```
I think it would be better if the openwrt.cmake file exports STAGING_DIR instead. openwrt.cmake is supposed to emulate the variables that are normally set by openwrt (in include/cmake.mk) so we can easily build outside of openwrt.

arnout commented on framework/cmake/Findsafec.cmake:
```diff
@@ -0,0 +1,35 @@
+# See LICENSE file for more details.
+
+find_library(SAFEC_LIBRARY "libsafec-3.3.so")
+find_path(SAFEC_INCLUDE_DIRS libsafec/safe_str_lib.h)
```
Unless I'm very mistaken, libsafec has a pkg-config file, so we can use pkg_check_modules instead (cfr. Findjson-c.cmake).

arnout commented on framework/cmake/Findsafec.cmake:
```diff
@@ -0,0 +1,35 @@
+
+    # Includes
+    set_target_properties(safec PROPERTIES
+        INTERFACE_INCLUDE_DIRECTORIES "${SAFEC_INCLUDE_DIRS}/libsafec"
```
I don't think this is correct, because the `#include` statements use `<libsafec/safe_str_lib.h>` so it should be just `${SAFEC_INCLUDE_DIRS}`. It happens to work because it's installed in staging/usr/include so ti's already in the default search path.

arnout commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -14,29 +14,64 @@ file(GLOB_RECURSE bpl_common_sources ${MODULE_PATH}/common/*.c*)
+        add_definitions(-DHAVE_C99)
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99")
+
+    elseif (TARGET_PLATFORM_TYPE STREQUAL "turris-omnia")
```
This approach makes no sense to me. We should have a BPL_TYPE that can be uci or linux (is brcm still used?), and set the BPL_TYPE based on TARGET_PLATFORM.

Also, I think Turris Omnia should also use uci bpl (eventually...).

tomereli replied:
brcm is not used. @vitalybu I suggest to also delete the whole brcm, fapi and wav bwl implementations as part of this PR (as you also removed bpl cal for older UGWs - this can be done in the same commit)

vitalybu replied:
BRCM is not used today, but it doesn't mean that we wouldn't port prplMesh to a Broadcom based platform sometime in the future. I would keep it for now.
Same goes for the WAV BWL... I know that we currently use DWPAL, but I didn't lose hope that at some point we'll be able to switch to the cleaner (and leaner) WAV implementation.
I'll remove the FAPI implementations.

tomereli replied:
We can always go back in history and bring them back. At this point in time there is just no need for it so I think we should remove it. It is basically dead code at the moment...

arnout commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -14,29 +14,64 @@ file(GLOB_RECURSE bpl_common_sources ${MODULE_PATH}/common/*.c*)
-    # UGW specific includes
-    include_directories(${MODULE_PATH}/ugw/cfg/cal)
+# OpenWRT
+if (TARGET_PLATFORM STREQUAL "openwrt")
```
Here you're changing the meaning of TARGET_PLATFORM, which is actually part of the following commit. There should be a first commit that changes the meaning of TARGET_PLATFORM. Also, it would be nicer if the part where support for ugw on bpl UCI is added is done in a separate commit. So three commits in total: one to change turris-omnia to openwrt; another one that changes rdkb to uci; a third one that adds ugw support to bpl.

arnout commented on common/beerocks/scripts/CMakeLists.txt:
```diff
@@ -2,9 +2,12 @@
-elseif (TARGET_PLATFORM STREQUAL "turris-omnia")
-    set(PLATFORM_DB_PATH "/opt/prplmesh/share")
+elseif (TARGET_PLATFORM STREQUAL "openwrt")
+    set(PLATFORM_DB_PATH "/opt/prplMesh/share")
```
If you change /opt/prplmesh to /opt/prplMesh, it should be done everywhere in a single commit.

But, as mentioned elsewhere, it actually shouldn't be done at all :-)

tomereli replied:
Agree, no need to change it at all in the first place

arnout commented on common/beerocks/scripts/CMakeLists.txt:
```diff
@@ -2,9 +2,12 @@
+    set(PLATFORM_DB_PATH "/opt/prplMesh/share")
     set(PLATFORM_DB_PATH_TEMP "/tmp/beerocks")
+    if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
+        set(EXTRA_LD_PATH "/opt/intel/lib")
```
It sucks that this is needed. Normally, cmake should set the rpath correctly so we should investigate why it doesn't happen here.

tomereli replied:
It is not needed. I opened a PR to fix it in UGW, please remove it from prplmesh. No need for the EXTRA_LD_PATH.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -163,13 +163,13 @@ stop_function() {
 
 main_agent_operational() {
-    pgrep -la beerocks_agent | grep -v wlan > /dev/null 2>&1 || return 1
+    pgrep -fx "@BEEROCKS_INSTALL_PATH@/bin/beerocks_agent" > /dev/null 2>&1 || return 1
```
The commit message should definitely explain why this change is needed...

vitalybu replied:
Updated the commit message

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -163,13 +163,13 @@ stop_function() {
-    pgrep -la beerocks_agent | grep $2 | grep -v grep > /dev/null 2>&1 || return 1
-    grep -q 'CSA' "$1/beerocks_agent_$2.log"
+    pgrep -fx "@BEEROCKS_INSTALL_PATH@/bin/beerocks_agent -i $2" > /dev/null 2>&1 || return 1
+    grep -q 'goto STATE_OPERATIONAL' "$1/beerocks_agent_$2.log"
```
I certainly agree that this change is needed (in fact, I though we had done this already a while ago), but it should be in a separate commit with a good commit message.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -18,7 +18,7 @@ using namespace net;
 agent_ucc_listener::agent_ucc_listener(uint16_t port, const std::string &vendor,
                                        const std::string &model, const std::string &bridge_iface,
-                                       Socket **controller_sd)
+                                       SocketClient **controller_sd)
```
Change is correct, but the commit message is not entirely correct: it's an upcast from `SocketClient*` to `Socket*` - the additional pointer is the problem.

arnout commented on CMakeLists.txt:
```diff
@@ -8,6 +8,9 @@ endif()
 option (BUILD_CONTROLLER "Build EasyMesh controller" ON)
 
+# Add global project cmake modules path
+list(INSERT CMAKE_MODULE_PATH 0 ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
```
Ideally, the same commit should also remove this from the four component CMakeLists.txt (and the commit should do only that), but that's just a nitpick.

arnout reviewed (changes requested):
The first commit should be split in two - which you can already see in the subject line, because it has "and" in it.

The second commit's commit message is not properly wrapped at 72 columns. Same for a few commits after it.

There are a few fundamental comments, a lot of nitpicks, and many commits that can be merged as is (with a properly wrapped commit message). I propose to make a separate PR with the stuff that is ready to merge. I can do that if you prefer.

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg_uci.c:
```diff
@@ -6,25 +6,10 @@
 # of the Intel Software License Agreement for the Intel(R) Cable and GW Software Development Kit"  #
 ##################################################################################################*/
 
-/*! \file 	uci_wrapper.c
```
By removing these lines you uncovered the fact that this file doesn't have the correct license :) So please update the license to our BSD+Patent which is used everywhere else. Same goes to the header file which is completely missing the license.

vitalybu replied:
Fixed

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg_uci.h:
```diff
@@ -0,0 +1,83 @@
+} // extern "C"
+#endif
+
+#endif // BPL_CFG_UCI_H_c
```
Missing line break

vitalybu replied:
Fixed

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg_uci.h:
```diff
@@ -0,0 +1,83 @@
+#ifndef BPL_CFG_UCI_H_
```
Missing license header

vitalybu replied:
Fixed

tomereli commented on framework/CMakeLists.txt:
```diff
@@ -64,8 +64,8 @@ if(TARGET_PLATFORM STREQUAL "linux")
-elseif (TARGET_PLATFORM STREQUAL "turris-omnia")
-    set(MAPF_ROOT "/opt/prplmesh")
+elseif (TARGET_PLATFORM STREQUAL "openwrt")
+    set(MAPF_ROOT "/opt/prplMesh")
```
Why not set `MAPF_ROOT` to be `/opt/prplmesh` by default and remove the dependency on the platform type?

```
set(MAPF_ROOT "/opt/prplmesh") # can be moved to multiap-helpers.cmake
set(TMP_PATH "/tmp/prplmesh")

if(TARGET_PLATFORM STREQUAL "linux")
    set(MAPF_ROOT "${CMAKE_INSTALL_PREFIX}")
    option(USE_USER_TMP_PATH "use /tmp/$USER instead of /tmp path" ON)
    if(USE_USER_TMP_PATH)
        set(USER "$ENV{USER}")
        set(TMP_PATH "/tmp/${USER}")
    endif()
endif()
```

tomereli commented on common/CMakeLists.txt:
```diff
@@ -15,15 +15,21 @@ include(GNUInstallDirs)
+if (TARGET_PLATFORM STREQUAL "openwrt")
+    # UGW
+    if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
+        add_definitions(-DBEEROCKS_UGW)
```
It's about time we get rid of BEEROCKS_UGW. it is only used on beerocks_version - so OK, for now, to keep it only for version print.

Other uses - 

- network_utils.cpp - used for endianess in `mac_to_uint64` and broken for none UGW anyway (need to remove BEEROCKS_BRCM support here)
- BWL DWPAL - used for different safec libraries includes. It would be better to have a different h file which will include the correct header based on the safec library found by the BWL's CMakeLists.txt
- BPL UCI - same as BWL comment

tomereli commented on common/CMakeLists.txt:
```diff
@@ -15,15 +15,21 @@ include(GNUInstallDirs)
+        add_definitions(-DBEEROCKS_UGW)
+    # Turris Omnia
+    elseif (TARGET_PLATFORM_TYPE STREQUAL "turris-omnia")
+        set(BEEROCKS_WLAN2_IFACE "wlan1")
```
I think its better if we require turris omnia platforms to rename the wlan1 to wlan2 then have it hardcoded in our build system.

arnout replied:
??? Why would you want to rename wlan1 to wlan2? I think it's better to require whatever crazy platform that has wlan0 and wlan2 to rename wlan2 to wlan1 :-)

tomereli replied:
Perhaps I'm biased towards Intel, but from what can remember we are not the only ones who mark STA interfaces with odd indexes. So wlan0,2,4 - APs, wlan1,3,5 - STAs.
Anyway, in most prplmesh supported platforms we will have this naming convention - UGW, RDKB, prplWrt - since they are Intel based.
I think the best solution is to just add it to the platform DB and never depend on the interface name. But this, of course, can be done in a separate PR, so (nitpick).

arnout replied:
> Perhaps I'm biased towards Intel, but from what can remember we are not the only ones who mark STA interfaces with odd indexes. So wlan0,2,4 - APs, wlan1,3,5 - STAs.

You're biased :-) That is *definitely* not how things should be done. If you have three radios, the first interface (=BSS) on each of them (STA or AP, doesn't matter) will be called wlan0, wlan1, wlan2. If you add a second interface, it will be called wlan0.1. In fact, the names are just names, hostapd and wpa_supplicant explicitly construct names like that. (wlan0 itself is a bit special because that name is constructed by the kernel, because the kernel will automatically add an interface to the phy when the driver is loaded, and that name is wlanN (typically, it actually depends on the driver).

Broadcom at least also follows this scheme, though the interfaces are called wl0, wl0.1 there.

I do suppose wlan0 and wlan1 are both on phy0 in the intel case, right?

> I think the best solution is to just add it to the platform DB and never depend on the interface name. But this, of course, can be done in a separate PR, so (nitpick).

Well, actually, it is only used to generate the config files. I think it's a bit extreme to have build options that are *only* used to generate a default config file...

tomereli reviewed (changes requested):
Reviewed again till commit common: bwl: CMake updates for UGW support

I will continue the review later

Commit:
common/master: Remove unknown IP address state

prplMesh does not rely on the IP address of connected stations.
This commit removes the STATE_CONNECTED_IP_UNKNOWN state.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
cmake: Add CMake color variables to the multiap-helpers.cmake file

The multiap-helpers.cmake file is included from the top CMake file.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

arnout committed:
treewide: use 'openwrt' instead of 'turris-omnia' for openwrt

The platform that was previously called 'Turris Omnia' really is the
OpenWrt platform. Thus, use 'openwrt' instead of 'turris-omnia'.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented:
Superceeded by #514
